ansys/dpf/core/scoping.py

```diff
@@ -349,8 +353,204 @@ def deep_copy(self, server=None):
+
+        Returns
+        -------
+        location : str
+            Location of the IDs.
+        """
+        return self._location
+
+    def _set_location(self, loc=locations.nodal):
+        """
+        Parameters
+        ----------
+        loc : str or core.locations enum
+            Location needed.
+
+        """
+        self._location = loc
+
+    @version_requires("2.1")
+    def _set_ids(self, ids):
+        """
+        Parameters
+        ----------
+        ids : list of int
+            IDs to set.
+
+        Notes
+        -----
+        Print a progress bar.
+        """
+        # must convert to a list for gRPC
+        if isinstance(ids, range):
+            ids = np.array(list(ids), dtype=np.int32)
+        elif not isinstance(ids, (np.ndarray, np.generic)):
+            ids = np.array(ids, dtype=np.int32)
+        else:
+            ids = np.array(list(ids), dtype=np.int32)
+
+        self._scoping_ids_copy = ids
+        self.__init_map__()
+
+    def _get_ids(self, np_array=False):
+        """
+        Returns
+        -------
+        ids : list[int], numpy.array (if np_array==True)
+            List of IDs.
+
+        Notes
+        -----
+        Print a progress bar.
+        """
+        if np_array:
+            return self._scoping_ids_copy
+        else:
+            return self._scoping_ids_copy.tolist()
+
+    def set_id(self, index, scopingid):
+        """Set the ID of a scoping's index.
+
+        Parameters
+        ----------
+        index : int
+            Index of the scoping.
+        scopingid : int
+            ID of the scoping.
+        """
+        if self._count() <= index:
+            self._scoping_ids_copy.resize(index+1)
+        self._scoping_ids_copy[index] = scopingid
+        self._mapper[scopingid] = index
+
+    def append(self, id):
+        self._scoping_ids_copy.append(id)
+        self._mapper[id] = len(self)-1
+
+    def _get_id(self, index):
+        """Retrieve the index that the scoping ID is located on.
+
+        Parameters
+        ----------
+        index : int
+            Index of the scoping
+
+        Returns
+        -------
+        id : int
+            ID of the scoping's index.
+        """
+        return self._scoping_ids_copy[index]
+
+    def _get_index(self, scopingid):
+        """Retrieve an ID corresponding to an ID in the scoping.
+
+        Parameters
+        ----------
+        id : int
+            ID to retrieve.
+
+        Returns
+        -------
+        index : int
+            Index of the ID.
+        """
+        return self._mapper[scopingid]
+
+    def release_data(self):
+        """Release the data."""
+        self.ids = self._scoping_ids_copy
+        super()._set_location(self._location)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, tb):
+        if tb is None:
+            self.release_data()
+        else:
+            print(tb)
+
+    def __del__(self):
```

Comment:
Consider updating the server on destruction.

Reply:
You're right, I did it